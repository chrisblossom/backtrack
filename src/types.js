/* @flow */

/* eslint-disable no-use-before-define */

type ShellCommand = string;
type CustomFn = Function;

export type NamedTask = {|
    name: string,
    task: Task,
|};

export type AllTaskTypes = ShellCommand | CustomFn | NamedTask;

export type Lifecycle = $ReadOnlyArray<
    AllTaskTypes | $ReadOnlyArray<AllTaskTypes>,
>;

export type Task = AllTaskTypes | Lifecycle;

export type Copy = {|
    dest: string,
    src: string,
    hash?: boolean,
|};

export type NormalizedClean = {|
    del: $ReadOnlyArray<string>,
    makeDirs: $ReadOnlyArray<string>,
    copy: $ReadOnlyArray<Copy>,
|};

export type Clean = {|
    del?: string | $ReadOnlyArray<string>,
    makeDirs?: string | $ReadOnlyArray<string>,
    copy?: Copy | $ReadOnlyArray<Copy>,
    [any]: empty,
|};

type CopyFileOptions = {|
    +makeDirs?: $ReadOnlyArray<string> | string,
    +skip?: $ReadOnlyArray<string> | string,
    +allowChanges?: $ReadOnlyArray<string> | string | boolean,
    +ignoreUpdates?: $ReadOnlyArray<string> | string | boolean,
|};

export type CopyFile = {|
    +src: string,
    +dest: string,
    +allowChanges?: boolean,
    +ignoreUpdates?: boolean,
|};

export type FileManager = $ReadOnlyArray<CopyFile | CopyFileOptions>;

// eslint-disable-next-line flowtype/require-exact-type
export type Config = {
    [key: string]: any,
};

type PresetTask =
    | AllTaskTypes
    | false
    | $ReadOnlyArray<AllTaskTypes | $ReadOnlyArray<AllTaskTypes> | false>;

export type Files =
    | CopyFile
    | CopyFileOptions
    | false
    | $ReadOnlyArray<CopyFile | CopyFileOptions | false>;

// eslint-disable-next-line flowtype/require-exact-type
export type Resolve = { [key: string]: string };

// eslint-disable-next-line flowtype/require-exact-type
export type Preset = {
    presets?: string | $ReadOnlyArray<string>,

    build?: PresetTask,
    dev?: PresetTask,
    lint?: PresetTask,
    format?: PresetTask,
    test?: PresetTask,

    clean?: Clean | false | $ReadOnlyArray<Clean | false>,
    files?: Files,
    packageJson?: PackageJson | false | $ReadOnlyArray<PackageJson | false>,
    config?: Config | false | $ReadOnlyArray<Config | false>,

    [key: string]: PresetTask,
};

// eslint-disable-next-line flowtype/require-exact-type
export type Lifecycles = {
    build?: Lifecycle,
    dev?: Lifecycle,
    lint?: Lifecycle,
    test?: Lifecycle,
    format?: Lifecycle,

    clean?: NormalizedClean,
    files?: ParsedFiles,
    packageJson?: $ReadOnlyArray<PackageJson>,
    config?: $ReadOnlyArray<Config>,
    resolve?: Resolve,

    [key: string]: Lifecycle,
};

export type ParsedFiles = {|
    src: {|
        files: $ReadOnlyArray<string>,
        absolute: { [key: string]: string },
        hash: { [key: string]: string },
    |},
    dest: {|
        files: $ReadOnlyArray<string>,
        absolute: { [key: string]: string },
        hash: { [key: string]: string },
        allowChanges: { [key: string]: boolean },
        ignoreUpdates: { [key: string]: boolean },
    |},
    makeDirs: $ReadOnlyArray<string>,
|};

export type FileStats = $ReadOnly<{ [key: string]: string }>;

export type DirStats = $ReadOnlyArray<string>;

export type FileManagerStats = $ReadOnly<{
    directories?: DirStats,
    files?: FileStats,
    [any]: empty,
}>;

export type StatsFile = $ReadOnly<{
    fileManager?: FileManagerStats,
    packageJson?: PackageJson,
    // https://github.com/facebook/flow/issues/4582
    // https://github.com/facebook/flow/issues/2386
    [any]: empty,
}>;

export type Scripts = $ReadOnly<{ [key: string]: string | null }>;

export type PackageJson = $ReadOnly<{
    scripts?: Scripts,
}>;

export type CustomError = {|
    +message: string,
    +exitCode: number,
|};
