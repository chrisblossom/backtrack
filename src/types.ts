/* eslint-disable no-use-before-define */

type ShellCommand = string;
type CustomFn = Function;

export type NamedTask = Readonly<{
	name: string;
	task: Task;
}>;

export type AllTaskTypes = ShellCommand | CustomFn | NamedTask;

export type Lifecycle = ReadonlyArray<
	AllTaskTypes | ReadonlyArray<AllTaskTypes>
>;

export type Task = AllTaskTypes | Lifecycle;

export type Copy = {
	dest: string;
	src: string;
	hash?: boolean;
};

export type NormalizedClean = {
	del: Array<string>;
	makeDirs: Array<string>;
	copy: Array<Copy>;
};

export type Clean = {
	del?: string | string[];
	makeDirs?: string | string[];
	copy?: Copy | Copy[];
};

export type CopyFileOptions = {
	makeDirs?: string[] | string;
	skip?: string[] | string;
	allowChanges?: string[] | string | boolean;
	ignoreUpdates?: string[] | string | boolean;
};

export function isCopyFileOptions(obj: any): obj is CopyFileOptions {
	return obj.src === undefined && obj.dest === undefined;
}

export type CopyFile = {
	src: string;
	dest: string;
	allowChanges?: boolean;
	ignoreUpdates?: boolean;
};

export function isCopyFile(obj: any): obj is CopyFile {
	return obj.src !== undefined && obj.dest !== undefined;
}

export type FileManager = Array<CopyFile | CopyFileOptions>;

export type Config = {
	[key: string]: any;
};

type PresetTask =
	| AllTaskTypes
	| false
	| ReadonlyArray<AllTaskTypes | ReadonlyArray<AllTaskTypes> | false>;

export type Files =
	| CopyFile
	| CopyFileOptions
	| false
	| ReadonlyArray<CopyFile | CopyFileOptions | false>;

export type Resolve = { [key: string]: string };

export type Preset = {
	presets?: string | ReadonlyArray<string | [string, {}]>;

	build?: PresetTask;
	dev?: PresetTask;
	lint?: PresetTask;
	format?: PresetTask;
	test?: PresetTask;

	clean?: Clean | false | ReadonlyArray<Clean | false>;
	files?: Files;
	packageJson?: PackageJson | false | ReadonlyArray<PackageJson | false>;
	config?: Config | false | ReadonlyArray<Config | false>;
} & { [key: string]: PresetTask };

export type Lifecycles = {
	build?: Lifecycle;
	dev?: Lifecycle;
	lint?: Lifecycle;
	test?: Lifecycle;
	format?: Lifecycle;

	clean?: NormalizedClean;
	files?: ParsedFiles;
	packageJson?: ReadonlyArray<PackageJson>;
	config?: ReadonlyArray<Config>;
	resolve?: Resolve;
} & { [key: string]: Lifecycle };

export type ParsedFiles = {
	src: {
		files: string[];
		absolute: { [key: string]: string };
		hash: { [key: string]: string };
	};
	dest: {
		files: string[];
		absolute: { [key: string]: string };
		hash: { [key: string]: string };
		allowChanges: { [key: string]: boolean };
		ignoreUpdates: { [key: string]: boolean };
	};
	makeDirs: string[];
};

export type FileStats = Readonly<{ [key: string]: string }>;

export type DirStats = ReadonlyArray<string>;

export type FileManagerStats = Readonly<{
	directories?: DirStats;
	files?: FileStats;
}>;

export type StatsFile = Readonly<{
	fileManager?: FileManagerStats;
	packageJson?: PackageJson;
}>;

export type Scripts = Readonly<{ [key: string]: string | null }>;

export type PackageJson = Readonly<{
	scripts?: Scripts;
}> &
	Readonly<{ [key: string]: unknown }>;

export type CustomError = Readonly<{
	message: string;
	exitCode: number;
}>;
