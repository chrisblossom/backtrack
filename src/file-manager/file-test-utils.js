/* @flow */

/* eslint-disable arrow-body-style */

import type { FileManager } from '../types.js';

const filesPreprocessor = (args) =>
    // $FlowIgnore
    require('../options-file/files-preprocessor').filesPreprocessor(args);
const filesParser = (args) =>
    // $FlowIgnore
    require('../options-file/files-post-processor').filesPostProcessor(args);
const getFileStats = (parsedFiles) =>
    require('./get-file-stats').getFileStats(parsedFiles);

function fileStats(previousFiles: FileManager) {
    const normalize = filesPreprocessor({ value: previousFiles });

    const parsedPreviousFiles = filesParser({ value: normalize });

    const previousStats = getFileStats(parsedPreviousFiles).files;

    return previousStats;
}

function fileInfo(files: FileManager = [], previousFiles: FileManager = []) {
    const previousStats = fileStats(previousFiles);
    const normalize = filesPreprocessor({ value: files });

    const parsedFiles = filesParser({ value: normalize });

    return {
        previousStats,
        parsedFiles,
    };
}

export { fileInfo, fileStats };
