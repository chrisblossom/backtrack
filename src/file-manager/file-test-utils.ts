/* eslint-disable arrow-body-style,@typescript-eslint/no-unsafe-member-access,@typescript-eslint/no-unsafe-return,@typescript-eslint/no-unsafe-call,@typescript-eslint/no-unsafe-assignment */

import { FileManager, FileManagerStats } from '../types';

const filesPreprocessor = (args: any) =>
	require('../options-file/files-preprocessor').filesPreprocessor(args);
const filesParser = (args: any) =>
	require('../options-file/files-post-processor').filesPostProcessor(args);
const getFileStats = (parsedFiles: any) =>
	require('./get-file-stats').getFileStats(parsedFiles);

function fileStats(previousFiles: FileManager): FileManagerStats {
	const normalize = filesPreprocessor({ value: previousFiles });

	const parsedPreviousFiles = filesParser({ value: normalize });

	const previousStats = getFileStats(parsedPreviousFiles).files;

	return previousStats;
}

interface FileInfo {
	previousStats: FileManagerStats;
	parsedFiles: FileManager;
}

function fileInfo(
	files: FileManager = [],
	previousFiles: FileManager = [],
): FileInfo {
	const previousStats = fileStats(previousFiles);
	const normalize = filesPreprocessor({ value: files });

	const parsedFiles = filesParser({ value: normalize });

	return {
		previousStats,
		parsedFiles,
	};
}

export { fileInfo, fileStats };
