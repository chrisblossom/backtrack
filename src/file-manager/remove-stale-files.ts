import path from 'path';
import { existsSync } from 'fs';
import del from 'del';
import { rootPath } from '../config/paths';
import log from '../utils/log';
import { getFileHash } from '../utils/get-file-hash';
import { backupFile } from '../utils/backup-file';
import { ParsedFiles, FileStats } from '../types';

async function backupChangedFile(removeFilePath: string): Promise<void> {
	const backupResult = await backupFile(removeFilePath);
	if (backupResult === undefined) {
		return;
	}

	const relativeBackupFilePath = path.relative(rootPath, backupResult.file);

	const relativeFilePath = path.relative(rootPath, removeFilePath);

	log.warn(
		`'${relativeFilePath}' has been modified and is no longer managed. Moving to '${relativeBackupFilePath}'`,
	);
}

async function removeStaleFiles(
	files: ParsedFiles,
	previousStats: FileStats = {},
): Promise<void> {
	const pending: Promise<unknown>[] = [];

	for (const relativeFile of Object.keys(previousStats)) {
		const previousFileHash = previousStats[relativeFile];

		/**
		 * Do nothing if dest file is still managed
		 */
		if (files.dest.files.includes(relativeFile) === false) {
			/**
			 * Any file in previousStats not found in files should be removed
			 */
			const removeFilePath = path.resolve(rootPath, relativeFile);

			/**
			 * Do nothing if file already has been removed
			 */
			const exists = existsSync(removeFilePath);
			if (exists === true) {
				const removeFileHash = getFileHash(removeFilePath);

				/**
				 * Remove file if file has not changed
				 */
				if (previousFileHash === removeFileHash) {
					log.info(`Removing file: ${relativeFile}`);

					const removeFile = del(removeFilePath);
					pending.push(removeFile);
				} else {
					/**
					 * Backup file if changed
					 */
					const backupFilePending = backupChangedFile(removeFilePath);

					pending.push(backupFilePending);
				}
			}
		}

		/**
		 * Handle -latest files for allowChanges
		 */
		const { base: filename, dir, ext = '' } = path.parse(relativeFile);
		const latestFilename = `${filename}-latest${ext}`;
		const latestFilePath = path.resolve(rootPath, dir, latestFilename);

		const latestFileExists = existsSync(latestFilePath);

		const destFileIndex = files.dest.files.indexOf(relativeFile);
		const srcFile = files.src.files[destFileIndex];
		const srcHash = files.src.hash[srcFile];
		const destHash = files.dest.hash[relativeFile];

		/**
		 * Remove -latest file if dest file is equal to source
		 */
		if (latestFileExists && srcHash === destHash) {
			const relativeLatestFile = path.relative(dir, latestFilename);
			log.info(`Removing file: ${relativeLatestFile}`);

			const removeFile = del(latestFilePath);
			pending.push(removeFile);
		}
	}

	await Promise.all(pending);
}

export { removeStaleFiles };
