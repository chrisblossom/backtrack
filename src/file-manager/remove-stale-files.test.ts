import path from 'path';
import { fileInfo } from './file-test-utils';
import { FileManager } from '../types';

const removeStaleFiles = (files: any, previousStats: any) =>
	require('./remove-stale-files').removeStaleFiles(files, previousStats);

describe('removeStaleFiles', () => {
	const cwd = process.cwd();
	let del: any;
	let move: any;

	beforeEach(() => {
		del = require.requireMock('del');
		jest.mock('del', () => jest.fn(() => Promise.resolve()));

		move = require.requireMock('fs-extra').move;
		jest.mock('fs-extra', () => ({
			move: jest.fn(() => Promise.resolve()),
		}));

		jest.mock('../utils/log', () => ({
			warn: jest.fn(),
			info: jest.fn(),
			error: jest.fn(),
			success: jest.fn(),
		}));
	});

	afterEach(() => {
		process.chdir(cwd);
	});

	test('does nothing if no stats file', async () => {
		const dir = path.resolve(__dirname, '__sandbox__/stats1/');
		process.chdir(dir);

		const files = [
			{
				src: path.resolve(dir, 'nested/other.js'),
				dest: 'nested/other.js',
			},
		];

		const { parsedFiles, previousStats } = fileInfo(files);

		await removeStaleFiles(parsedFiles, previousStats);

		expect(del.mock.calls).toEqual([]);
		expect(move.mock.calls).toEqual([]);
	});

	test('does nothing if all files matched', async () => {
		const dir = path.resolve(__dirname, '__sandbox__/stats1/');
		process.chdir(dir);

		const previousFiles = [
			{
				src: path.resolve(dir, 'nested/file1.js'),
				dest: 'nested/file1.js',
			},
			{
				src: path.resolve(dir, 'nested/other.js'),
				dest: 'nested/other.js',
			},
		];

		const files = previousFiles;

		const { parsedFiles, previousStats } = fileInfo(files, previousFiles);

		await removeStaleFiles(parsedFiles, previousStats);

		expect(del.mock.calls).toEqual([]);
		expect(move.mock.calls).toEqual([]);
	});

	test('does nothing if files added', async () => {
		const dir = path.resolve(__dirname, '__sandbox__/stats1/');
		process.chdir(dir);

		const previousFiles = [
			{
				src: path.resolve(dir, 'nested/file1.js'),
				dest: 'nested/file1.js',
			},
		];

		const files = [
			{
				src: path.resolve(dir, 'nested/file1.js'),
				dest: 'nested/file1.js',
			},
			{
				src: path.resolve(dir, 'nested/other.js'),
				dest: 'nested/other.js',
			},
		];

		const { parsedFiles, previousStats } = fileInfo(files, previousFiles);

		await removeStaleFiles(parsedFiles, previousStats);

		expect(del.mock.calls).toEqual([]);
		expect(move.mock.calls).toEqual([]);
	});

	test('removes stale files', async () => {
		const dir = path.resolve(__dirname, '__sandbox__/stats1/');
		process.chdir(dir);

		const previousFiles = [
			{
				src: path.resolve(dir, 'nested/file1.js'),
				dest: 'nested/file1.js',
			},
			{
				src: path.resolve(dir, 'nested/inside.js'),
				dest: 'nested/inside.js',
			},
		];

		const files = [
			{
				src: path.resolve(dir, 'nested/file1.js'),
				dest: 'nested/file1.js',
			},
		];

		const { parsedFiles, previousStats } = fileInfo(files, previousFiles);

		await removeStaleFiles(parsedFiles, previousStats);

		expect(del.mock.calls).toMatchSnapshot();
		expect(move.mock.calls).toEqual([]);
	});

	test('backup file if removed and destination hash different', async () => {
		const dir = path.resolve(__dirname, '__sandbox__/stats1/');
		process.chdir(dir);

		const files: FileManager = [];

		const previousStats = {
			'nested/file1.js': 'fake',
		};

		const { parsedFiles } = fileInfo(files);

		await removeStaleFiles(parsedFiles, previousStats);

		expect(move.mock.calls).toMatchSnapshot();
		expect(del.mock.calls).toEqual([]);
	});

	test('allowChanges - removes -latest file when source and dest hash match', async () => {
		const dir = path.resolve(__dirname, '__sandbox__/stats1/');
		process.chdir(dir);

		const previousFiles = [
			{
				src: path.resolve(dir, 'file2.js'),
				dest: 'file2.js',
			},
		];

		const files = previousFiles;

		const { parsedFiles, previousStats } = fileInfo(files, previousFiles);

		await removeStaleFiles(parsedFiles, previousStats);

		expect(del.mock.calls).toMatchSnapshot();
		expect(move.mock.calls).toEqual([]);
	});

	test('allowChanges - removes -latest file when dest is no longer managed', async () => {
		const dir = path.resolve(__dirname, '__sandbox__/stats1/');
		process.chdir(dir);

		const previousFiles = [
			{
				src: path.resolve(dir, 'file2.js'),
				dest: 'file2.js',
			},
		];

		const files: FileManager = [];

		const { parsedFiles, previousStats } = fileInfo(files, previousFiles);

		await removeStaleFiles(parsedFiles, previousStats);

		expect(del.mock.calls).toMatchSnapshot();
		expect(move.mock.calls).toEqual([]);
	});
});
