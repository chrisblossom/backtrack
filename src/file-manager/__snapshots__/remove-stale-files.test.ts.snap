// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`removeStaleFiles allowChanges - removes -latest file when dest is no longer managed 1`] = `
[
  [
    "<PROJECT_ROOT>/file2.js",
  ],
  [
    "<PROJECT_ROOT>/file2.js-latest.js",
  ],
]
`;

exports[`removeStaleFiles allowChanges - removes -latest file when source and dest hash match 1`] = `
[
  [
    "<PROJECT_ROOT>/file2.js-latest.js",
  ],
]
`;

exports[`removeStaleFiles backup file if removed and destination hash different 1`] = `
[
  [
    "<PROJECT_ROOT>/nested/file1.js",
    "<PROJECT_ROOT>/nested/file1.js-backup-971eb09e.js",
    {
      "overwrite": false,
    },
  ],
]
`;

exports[`removeStaleFiles removes stale files 1`] = `
[
  [
    "<PROJECT_ROOT>/nested/inside.js",
  ],
]
`;
